homeassistant/components/ssdp/__init__.py

```diff
@@ -141,38 +157,52 @@ def _async_headers_match(
 
 
 class Scanner:
-    """Class to manage SSDP scanning."""
+    """Class to manage SSDP searching and SSDP advertisements."""
 
-    def __init__(
-        self, hass: HomeAssistant, integration_matchers: dict[str, list[dict[str, str]]]
-    ) -> None:
+    def __init__(self, hass: HomeAssistant) -> None:
         """Initialize class."""
         self.hass = hass
-        self.seen: set[tuple[str, str | None]] = set()
-        self.cache: dict[tuple[str, str], Mapping[str, str]] = {}
-        self._integration_matchers = integration_matchers
         self._cancel_scan: Callable[[], None] | None = None
-        self._ssdp_listeners: list[SSDPListener] = []
-        self._callbacks: list[tuple[Callable[[dict], None], dict[str, str]]] = []
-        self.flow_dispatcher: FlowDispatcher | None = None
-        self.description_manager: DescriptionManager | None = None
+        self._ssdp_listeners: list[SsdpListener] = []
+        self._callbacks: list[tuple[SsdpCallback, dict[str, str]]] = []
+        self._flow_dispatcher: FlowDispatcher | None = None
+        self._description_cache: DescriptionCache | None = None
+        self._integration_matchers: dict[str, list[dict[str, str]]] | None = None
+
+    @property
+    def _ssdp_devices(self) -> set[SsdpDevice]:
+        """Get all seen devices."""
+        ssdp_devices: set[SsdpDevice] = set()
```

Comment:
Is `SsdpDevice` hashable in a way that the same device seen by different listeners will be considered the same?

Reply:
Currently not, should it be? I'm hesitant, it raises questions like which one to use for example. If a device is seen via multiple networks what would one want? Options:
- Get both (current functionality)
- Get any of the duplicates
- Get the last seen of the duplicates. Aalthough there probably isn't much difference in these, as searches repeat every minute on all interfaces. Any advertisements will update the timestamp. A `BYEBYE` causes the device to be deleted in the listener.

The only situation I can think of is where somebody plugs the device into another switch and the device is seen via another listener. In that case it'll have a newer timestamp and it would be best to use that new device. What do you think?

Reply:
With zeroconf its possible that a device would announce a different addressed based on which interface its broadcasting on.  I'm not sure that would be compliant or not in this instance but I wouldn't be surprised if it does happen in the wild.

I think what's confusing me here is the use of a `set` and not a `list`.  Is there a case where there would be multiple objects that are the same added to the set?

Reply:
Would something like

```
    @property
    def _ssdp_devices(self) -> set[SsdpDevice]:
        """Get all seen devices."""
        return itertools.from_iterable(ssdp_listener.devices.values() for ssdp_listener in self._ssdp_listeners)
```

make sense here?

Reply:
Agreed, the use of `set` is confusing here/doesn't provide any value. My idea was that there is no order in the devices, as the `list` might imply. I'll make it a `list`.